### PR TITLE
Replace unsound struct-cast simplification

### PR DESCRIPTION
**Pull request by smowton**
Previously the simplifier would simplify ((struct A)x) into (x) if x.type() and A
were structurally compatible (i.e. had the same named and typed fields). However this
isn't really a sound simplification, as it changes the type of the expression. I therefore
replace the rule with one that only applies within the common-case member expression, i.e.
simplifiying ((struct A)x).y -> x.y, which does not change the type of the outer expression.

Notes: @allredj @romainbrenguier initially I simply disabled the unsound simplification, however this caused two string-smoke-tests to fail because a hard `java.lang.String` -> `java.lang.CharSequence` cast went unsimplified, preventing constant propagation from completing and causing the `contains` and `append_string` tests to fail because they needed to use their general case rather than the special case where a parameter is known to be a constant. If you try disabling the new simplification introduced here you should be able to reproduce the problem -- this might be an interesting bug, or could just be a known existing weakness in the solver.

This solution could be generalised by permitting this simplification *only for inner expressions*, i.e. casts which do not occur at the root of the expression tree. @tautschnig do we want to try that, or is handling this special case enough?

### CODE REVIEW

**Comment by tautschnig** [comment]
We had that discussion elsewhere - would you happen to have pointers to that?

File: src/util/simplify_expr_struct.cpp [hunk 1]
```diff
@@ -208,6 +209,17 @@ bool simplify_exprt::simplify_member(exprt &expr)
...
     }
   }
+  else if(op.id() == ID_typecast)
+  {
+    // Try to look through member(cast(x)) if the cast is between structurally
+    // identical types:
+    if(base_type_eq(op_type, op.op0().type(), ns))
+    {
+      expr.op0() = op.op0();
+      simplify_rec(expr);
```
Comment (tautschnig) [inline src/util/simplify_expr_struct.cpp]: I don't think it's necessary to simplify recursively, I'd say a call to `simplify_member` should do the trick.

**Comment by tautschnig** [comment]
It would be great to have some regression tests that show that the previous one was unsound. Otherwise I'm fine with this special case.

**Comment by smowton** [comment]
@tautschnig for reference the previous discussion of the same problem is here: https://github.com/diffblue/cbmc/pull/1499#issuecomment-346793910

File: src/util/simplify_expr_struct.cpp (hunk 1 above)
Comment (tautschnig) [reply src/util/simplify_expr_struct.cpp]: I think this is the only bit keeping me from approving.

**Comment by smowton** [comment]
@tautschnig the string smoke tests were exposing the problem, so I've added an extra commit to verify they don't any longer. Tagging string developers to check this is reasonable.

I've also implemented your comment above.

File: regression/strings-smoke-tests/java_append_object/test.desc [hunk 2]
```diff
@@ -1,8 +1,9 @@
 FUTURE
 test_append_object.class
---refine-strings --string-max-length 1000
+--refine-strings --verbosity 10 --string-max-length 1000
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 Issue: diffblue/test-gen#82
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_append_object/test.desc]: There is a problem in the original test description file: this comment "Issue: ..." should have a preceding `--` line to separate strings-not-to-be-found-in-output from comments.

File: regression/strings-smoke-tests/java_char_array_init/test.desc [hunk 3]
```diff
@@ -1,8 +1,9 @@
 FUTURE
 test_init.class
---refine-strings --string-max-length 1000
+--refine-strings --verbosity 10 --string-max-length 1000
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 cbmc/test-gen#259
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_char_array_init/test.desc]: As above, the comment needs a preceding `--` line.

File: regression/strings-smoke-tests/java_equal/test_2.desc [hunk 4]
```diff
@@ -1,7 +1,8 @@
 KNOWNBUG
 test_equal_2.class
---refine-strings --string-max-length 100 --string-max-length 100
+--refine-strings --verbosity 10 --string-max-length 100 --string-max-length 100
 ^EXIT=0$
 ^SIGNAL=0$
 --
+non equal types
 https://github.com/diffblue/test-gen/issues/856
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_equal/test_2.desc]: As above: `--` required.

File: regression/strings-smoke-tests/java_if/test.desc [hunk 5]
```diff
@@ -1,9 +1,10 @@
...
 test.class
---refine-strings --string-max-length 100
+--refine-strings --verbosity 10 --string-max-length 100
 ^EXIT=10$
 ^SIGNAL=0$
 ^\[.*assertion.1\].* line 11.* SUCCESS$
 ^\[.*assertion.2\].* line 13.* FAILURE$
 --
+non equal types
 $ignoring\s*char\s*array
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_if/test.desc]: Why does this pattern start with a `$`?

File: regression/strings-smoke-tests/java_index_of_char/test.desc [hunk 6]
```diff
@@ -1,8 +1,9 @@
 CORE
 test_index_of_char.class
---refine-strings --string-max-length 1000
+--refine-strings --verbosity 10 --string-max-length 1000
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 Issue: cbmc/test-gen#77
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_index_of_char/test.desc]: As above: `--` required.

File: regression/strings-smoke-tests/java_int_to_string/test2_bug.desc [hunk 7]
```diff
@@ -1,11 +1,12 @@
...
+--refine-strings --verbosity 10
 ^EXIT=10$
 ^SIGNAL=0$
 assertion.* line 7 .* SUCCESS$
 assertion.* line 10 .* FAILURE$
 --
+non equal types
 --
 When string-max-length is not set, the solver can run out of memory in
 an unpredictable way.
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_int_to_string/test2_bug.desc]: Missing newline at end-of-file.

File: regression/strings-smoke-tests/java_length2/test.desc [hunk 8]
```diff
@@ -1,10 +1,11 @@
 FUTURE
 test.class
---refine-strings --string-max-length 1000 --function test.check
+--refine-strings --verbosity 10 --string-max-length 1000 --function test.check
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 This test currently fails because we do not handle the case where the argument
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_length2/test.desc]: As above: `--` needed

File: regression/strings-smoke-tests/java_replace/test.desc [hunk 9]
```diff
@@ -1,10 +1,11 @@
 FUTURE
 test_replace.class
---refine-strings --string-max-length 100
+--refine-strings --verbosity 10 --string-max-length 100
 ^EXIT=0$
 ^SIGNAL=0$
 ^VERIFICATION SUCCESSFUL$
 --
+non equal types
 diffblue/test-gen#256
```
Comment (tautschnig) [inline regression/strings-smoke-tests/java_replace/test.desc]: As above: `--` required

**Review by tautschnig — APPROVED** [review]
Looks good, but the *.desc files need some (mostly trivial) work.

**Comment by smowton** [comment]
@tautschnig changes addressed. Looks like my script to add the rejected lines wasn't quite smart enough :)

**Review by romainbrenguier — APPROVED** [review]
this looks fine

**Pull request merged**